### PR TITLE
86 90 92 small fixes

### LINKED ISSUE #86: Move file path defences in `UrbanCentre` class to `__init__`
**Description of the Feature to be Added**

Explore possible to move file path defensive checks earlier within the class to the `__init__()`. 

This snippet offers file path defences:

https://github.com/datasciencecampus/transport-network-performance/blob/66663568efcd7aa7cd552f25f37c4c1302e6555d/src/transport_performance/urban_centres/raster_uc.py#L129-L132

But it maybe more beneficial to the user to get errors when instantiating the `UrbanCentre` class as opposed to needing to call `get_urban_centres()` to trigger the error (which would then need the whole class to be re-instantiated).

This request relates to #84 and could be resolved at the same time. It also relates to the review of [PR #37](https://github.com/datasciencecampus/transport-network-performance/pull/37).

### LINKED ISSUE #90: `raster_uc.py` sometimes raises `TypeErrors` with mis matching variable names
**Description of the Feature to be Added**

Very minor point, but some `TypeErrors` messages don't match with the variable name, e.g.:

https://github.com/datasciencecampus/transport-network-performance/blob/66663568efcd7aa7cd552f25f37c4c1302e6555d/src/transport_performance/urban_centres/raster_uc.py#L276-L280

Could be useful to revisit the type error and make sure they are consistent with the variable names to help user debugging.

This request relates to the review of [PR #37](https://github.com/datasciencecampus/transport-network-performance/pull/37).

Additionally, the variable name `cell_fill_treshold` should be changed to `cell_fill_threshold` to correct this typo (requires an update of the unit tests too). This was picked up as part of PR #104.

### LINKED ISSUE #92: Allow `centre` tuple to be in the same CRS as the raster data
**Description of the Feature to be Added**

In `UrbanCentres`, it's clear in the docstring that `centre` should be in EPSG:4326 lat/long format, but it could be useful to allow users to define this in the same CRS as the raster data (or any CRS and then convert only if needed).

An example of this is in the `rasterpop` module, with the `aoi_bounds` and. `urban_centre_bounds` arguments:
https://github.com/datasciencecampus/transport-network-performance/blob/66663568efcd7aa7cd552f25f37c4c1302e6555d/src/transport_performance/population/rasterpop.py#L72-L81

They have respective `_crs` arguments that are by defualt `None`, meaning the CRS is the same as the rater. Otherwise, the provided CRS is used to covert the argument to the CRS of the population raster data.

An additional feature could be to use the centre of the BBOX as the default centroid in the case where no `centre` argument is provided.

This request relates to the review of [PR #37](https://github.com/datasciencecampus/transport-network-performance/pull/37).

### PR DESCRIPTION
**Pull request by SergioRec**
## Description
Fixes #86 
Fixes #90 
Fixes #92

## Motivation and Context
Small fixes to urban centre module and required changes to unit tests and example notebook. Small modifications to docstrings and error messages.

Changes are:
#86: defence for file changed from `_handle_path_like` to `_is_expected_filetype`. Added tests to account for different error outputs. Note that the output for `_is_expected_filetype` is None so this is not assigned now.
#90: revised error messages for consistency.
#92: by default, the assumed crs of the centre will be the same as the raster. If different, the argument `centre_crs` will need to be specified. Updated unit testing to check this new argument.

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?
- Code run locally.
- Unit tests run locally.
- Example notebook runs and produces expected outputs.

Test configuration details:
* OS: MacOS Ventura 13.4.1
* Python version: 3.9.13
* Java version: n/a
* Python management system: conda/pip

## Checklist:

- [x] My code follows the intended structure of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

## Additional comments
I've merged with dev to include changes from last week.

### CODE REVIEW

File: src/transport_performance/urban_centres/raster_uc.py [hunk 1]
```diff
@@ -212,7 +224,7 @@ def _cluster_cells(
         """
         if not isinstance(flag_array, np.ndarray):
             raise TypeError(
-                "`masked_rst` expected numpy array, "
+                "`flag_array` expected numpy array, "
```
Comment (cbrowning04) [inline src/transport_performance/urban_centres/raster_uc.py:227]: I would argue that this is an unclear error message as the `flag_array` variable is created within the function, therefore the user is not aware of its existence. They would only be aware of the `masked_rst` variable used to create `flag_arraay`

File: src/transport_performance/urban_centres/raster_uc.py [hunk 2]
```diff
@@ -352,12 +363,12 @@ def _fill_gaps(
                 "`urban_centres` expected numpy array, "
                 f"got {type(urban_centres).__name__}."
             )
-        if not isinstance(threshold, int):
+        if not isinstance(cell_fill_threshold, int):
             raise TypeError(
                 "`threshold` expected integer, "
-                f"got {type(threshold).__name__}"
+                f"got {type(cell_fill_threshold).__name__}"
             )
```
Comment (cbrowning04) [inline src/transport_performance/urban_centres/raster_uc.py]: This error message inconsistent with the param name. Error raised mentions `threshold`, param name is cell_fill_threshold.

File: src/transport_performance/urban_centres/raster_uc.py [hunk 3]
```diff
@@ -352,12 +363,12 @@ def _fill_gaps(
...
-        if not (5 <= threshold <= 8):
+        if not (5 <= cell_fill_threshold <= 8):
             raise ValueError(
                 "Wrong value for `threshold`, "
                 "please enter value between 5 and 8"
```
Comment (cbrowning04) [inline src/transport_performance/urban_centres/raster_uc.py]: The same thing as the above comment applies here

File: src/transport_performance/urban_centres/raster_uc.py [hunk 4]
```diff
@@ -453,24 +473,28 @@ def _vectorize_uc(
             )
         if not isinstance(aff, affine.Affine):
             raise TypeError("`aff` must be a valid Affine object")
-        if not isinstance(crs, rasterio.crs.CRS):
+        if not isinstance(raster_crs, rasterio.crs.CRS):
             raise TypeError("`crs` must be a valid rasterio.crs.CRS object")
```
Comment (cbrowning04) [inline src/transport_performance/urban_centres/raster_uc.py]: This error message does not align with the parameter name

**Review by cbrowning04 — CHANGES_REQUESTED** [review]
Hi @SergioRec . Great work applying these changes. All new functionality works as expected (e.g., centre_crs).  
I have a few small comments that shouldn't be too difficult to fix up.  

Finally, in issue 92, it is mentioned that the centroid of the passed bbox could be used as a centre if no argument is passed for centre. This doesn't appear to be implemented here.

File: src/transport_performance/urban_centres/raster_uc.py (hunk 1 above)
Comment (SergioRec) [reply src/transport_performance/urban_centres/raster_uc.py:227]: I agree with you, and I thought about what would be the best way to do this. The user won't be able to see `masked_rst` either, so my rationale is that if there's an error, the traceback would point to `flag_array` as `masked_rst` belongs to a different function. In this particular case, that function input doesn't come from the user so the defence is only useful for debugging.

File: src/transport_performance/urban_centres/raster_uc.py (hunk 4 above)
Comment (SergioRec) [reply src/transport_performance/urban_centres/raster_uc.py]: Corrected. I changed the name of the parameter after checking consistency and forgot to re-check.

File: src/transport_performance/urban_centres/raster_uc.py (hunk 3 above)
Comment (SergioRec) [reply src/transport_performance/urban_centres/raster_uc.py]: Corrected. I changed the name of the parameter after checking consistency and forgot to re-check.

File: src/transport_performance/urban_centres/raster_uc.py (hunk 2 above)
Comment (SergioRec) [reply src/transport_performance/urban_centres/raster_uc.py]: Corrected. I changed the name of the parameter after checking consistency and forgot to re-check.

**Comment by SergioRec** [comment]
I had missed the bit in #92 that mentioned inferring the centre from the bbox. This will require some refactoring:
- The function using the parameter will become too complex if adding new stuff (including more defences), so I will need to split it and/or create a new function to get the centre.
- There's a risk of the centroid of the bbox not including the cluster. We would need to find a way of going around this, or to give some understandable feedback to the user if this happens.
- New set of unit tests.

I'll create a new item for this.

I've corrected all the other issues you raised, so if you're happy with it this PR could be merged.

File: src/transport_performance/urban_centres/raster_uc.py (hunk 1 above)
Comment (cbrowning04) [reply src/transport_performance/urban_centres/raster_uc.py:227]: I'm happy to mark this as resolved then @SergioRec . If the user has no influence over this error, then it is simply an error used for debugging. The raised message is not as important in the event that the user has no influence over the error raised, therefore this message is fine.

File: src/transport_performance/urban_centres/raster_uc.py (hunk 2 above)
Comment (cbrowning04) [reply src/transport_performance/urban_centres/raster_uc.py]: Resolving as fixed

File: src/transport_performance/urban_centres/raster_uc.py (hunk 4 above)
Comment (cbrowning04) [reply src/transport_performance/urban_centres/raster_uc.py]: Resolving as fixed

File: src/transport_performance/urban_centres/raster_uc.py (hunk 3 above)
Comment (cbrowning04) [reply src/transport_performance/urban_centres/raster_uc.py]: Resolving as fixed

**Comment by cbrowning04** [comment]
> I had missed the bit in #92 that mentioned inferring the centre from the bbox. This will require some refactoring:
> 
>     * The function using the parameter will become too complex if adding new stuff (including more defences), so I will need to split it and/or create a new function to get the centre.
> 
>     * There's a risk of the centroid of the bbox not including the cluster. We would need to find a way of going around this, or to give some understandable feedback to the user if this happens.
> 
>     * New set of unit tests.
> 
> 
> I'll create a new item for this.
> 
> I've corrected all the other issues you raised, so if you're happy with it this PR could be merged.

Great thanks for this @SergioRec . Thanks for clearing up the stuff with #92 as well. Happy to merge this now.

**Review by cbrowning04 — APPROVED** [review]

**Pull request merged**